In [1]:
# Title: "Create and save exploratory data analysis figures of coffee quality dataset"
# Authors: "Arlin Cherian, Kristin Bunyan, Michelle Wang, Berkay Bulut"
# date: 2021-11-19

In [2]:
# loading packages and set up

import numpy as np
import pandas as pd
import altair as alt
from sklearn import datasets
from sklearn.model_selection import train_test_split

In [3]:
# Loading data set
coffee_df = pd.read_csv('https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-07-07/coffee_ratings.csv')
# check number of rows and columns
coffee_df.shape
coffee_df = coffee_df.query("species == 'Arabica'")
coffee_df

total_cup_points  species                       owner country_of_origin  \
0                90.58  Arabica                   metad plc          Ethiopia   
1                89.92  Arabica                   metad plc          Ethiopia   
2                89.75  Arabica    grounds for health admin         Guatemala   
3                89.00  Arabica         yidnekachew dabessa          Ethiopia   
4                88.83  Arabica                   metad plc          Ethiopia   
...                ...      ...                         ...               ...   
1306             68.33  Arabica    juan carlos garcia lopez            Mexico   
1307             67.92  Arabica     myriam kaplan-pasternak             Haiti   
1308             63.08  Arabica  exportadora atlantic, s.a.         Nicaragua   
1309             59.83  Arabica   juan luis alvarado romero         Guatemala   
1310              0.00  Arabica             bismarck castro          Honduras   

                                     farm_name                  lot_number  \
0                                    metad plc                         NaN   
1                                    metad plc                         NaN   
2     san marcos barrancas "san cristobal cuch                         NaN   
3        yidnekachew dabessa coffee plantation                         NaN   
4                                    metad plc                         NaN   
...                                        ...                         ...   
1306                             el centenario                         NaN   
1307                                 200 farms                         NaN   
1308                          finca las marías  017-053-0211/ 017-053-0212   
1309                            finca el limon                         NaN   
1310                              los hicaques                         103   

                                                   mill  \
0                                             metad plc   
1                                             metad plc   
2                                                   NaN   
3                                               wolensu   
4                                             metad plc   
...                                                 ...   
1306  la esperanza, municipio juchique de ferrer, ve...   
1307        coeb koperativ ekselsyo basen (350 members)   
1308                         beneficio atlantic condega   
1309                                   beneficio serben   
1310                                 cigrah s.a de c.v.   

                      ico_number                                company  \
0                      2014/2015      metad agricultural developmet plc   
1                      2014/2015      metad agricultural developmet plc   
2                            NaN                                    NaN   
3                            NaN  yidnekachew debessa coffee plantation   
4                      2014/2015      metad agricultural developmet plc   
...                          ...                                    ...   
1306                  1104328663                              terra mia   
1307                         NaN                           haiti coffee   
1308  017-053-0211/ 017-053-0212               exportadora atlantic s.a   
1309                  11/853/165                                unicafe   
1310                  13-111-053                      cigrah s.a de c.v   

           altitude  ...       color category_two_defects  \
0         1950-2200  ...       Green                    0   
1         1950-2200  ...       Green                    1   
2     1600 - 1800 m  ...         NaN                    0   
3         1800-2200  ...       Green                    2   
4         1950-2200  ...       Green                    2   
...             ...  ...         ...                  ...   
1306            900  ...        None                   20  

In [4]:
# Explore the structure of the dataset
coffee_df.info()

# Explore numerical features in the data set
coffee_df.describe()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1311 entries, 0 to 1310
Data columns (total 43 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   total_cup_points       1311 non-null   float64
 1   species                1311 non-null   object 
 2   owner                  1304 non-null   object 
 3   country_of_origin      1310 non-null   object 
 4   farm_name              955 non-null    object 
 5   lot_number             270 non-null    object 
 6   mill                   1001 non-null   object 
 7   ico_number             1165 non-null   object 
 8   company                1102 non-null   object 
 9   altitude               1088 non-null   object 
 10  region                 1254 non-null   object 
 11  producer               1081 non-null   object 
 12  number_of_bags         1311 non-null   int64  
 13  bag_weight             1311 non-null   object 
 14  in_country_partner     1311 non-null   object 
 15  harv

total_cup_points  number_of_bags        aroma       flavor  \
count       1311.000000     1311.000000  1311.000000  1311.000000   
mean          82.115927      153.887872     7.563806     7.518070   
std            3.515761      129.733734     0.378666     0.399979   
min            0.000000        0.000000     0.000000     0.000000   
25%           81.170000       14.500000     7.420000     7.330000   
50%           82.500000      175.000000     7.580000     7.580000   
75%           83.670000      275.000000     7.750000     7.750000   
max           90.580000     1062.000000     8.750000     8.830000   

        aftertaste      acidity         body      balance   uniformity  \
count  1311.000000  1311.000000  1311.000000  1311.000000  1311.000000   
mean      7.397696     7.533112     7.517727     7.517506     9.833394   
std       0.405119     0.381599     0.359213     0.406316     0.559343   
min       0.000000     0.000000     0.000000     0.000000     0.000000   
25%       7.250000     7.330000     7.330000     7.330000    10.000000   
50%       7.420000     7.500000     7.500000     7.500000    10.000000   
75%       7.580000     7.750000     7.670000     7.750000    10.000000   
max       8.670000     8.750000     8.580000     8.750000    10.000000   

        clean_cup    sweetness  cupper_points     moisture  \
count  1311.00000  1311.000000    1311.000000  1311.000000   
mean      9.83312     9.903272       7.497864     0.088863   
std       0.77135     0.530832       0.474610     0.047957   
min       0.00000     0.000000       0.000000     0.000000   
25%      10.00000    10.000000       7.250000     0.090000   
50%      10.00000    10.000000       7.500000     0.110000   
75%      10.00000    10.000000       7.750000     0.120000   
max      10.00000    10.000000      10.000000     0.280000   

       category_one_defects      quakers  category_two_defects  \
count           1311.000000  1310.000000           1311.000000   
mean               0.426392     0.177099              3.591915   
std                1.832415     0.840583              5.350371   
min                0.000000     0.000000              0.000000   
25%                0.000000     0.000000              0.000000   
50%                0.000000     0.000000              2.000000   
75%                0.000000     0.000000              4.000000   
max               31.000000    11.000000             55.000000   

       altitude_low_meters  altitude_high_meters  altitude_mean_meters  
count          1084.000000           1084.000000           1084.000000  
mean           1759.548954           1808.843803           1784.196379  
std            8767.847252           8767.187498           8767.016913  
min               1.000000              1.000000              1.000000  
25%            1100.000000           1100.000000           1100.000000  
50%            1310.640000           1350.000000           1310.640000  
75%            1600.000000           1650.000000           1600.000000  
max          190164.000000         190164.000000         190164.000000

According to the dataset decription our target, `total_cup_points` is a point scale from 0-100 with mean scale rating of 82.1 points, minimum of 0.0 and maximum of 90.6 points.

In [5]:
# How many unique country of origin of coffee beans are included in the dataset?
print(coffee_df['country_of_origin'].nunique())
print(coffee_df['country_of_origin'].unique())

36
['Ethiopia' 'Guatemala' 'Brazil' 'Peru' 'United States'
 'United States (Hawaii)' 'Indonesia' 'China' 'Costa Rica' 'Mexico'
 'Uganda' 'Honduras' 'Taiwan' 'Nicaragua' 'Tanzania, United Republic Of'
 'Kenya' 'Thailand' 'Colombia' 'Panama' 'Papua New Guinea' 'El Salvador'
 'Japan' 'Ecuador' 'United States (Puerto Rico)' 'Haiti' 'Burundi'
 'Vietnam' 'Philippines' 'Rwanda' 'Malawi' 'Laos' 'Zambia' 'Myanmar'
 'Mauritius' 'Cote d?Ivoire' nan 'India']


In [6]:
# What are some regions of coffee production are included in the dataset?
print(coffee_df['region'].nunique())
print(coffee_df['region'].unique())

343
['guji-hambela' nan 'oromia' 'oromiya' 'snnp/kaffa zone,gimbowereda'
 'antioquia' 'kona' 'sulawesi' 'yirgacheffe' 'yunnan' 'gedio' 'san ramon'
 'xalapa' 'sidamo' 'south of minas' 'kapchorwa eastern' 'comayagua'
 'leye, alishan township, chiayi county' 'vale da grama'
 'west and central valley' 'muranga' 'chiang rai'
 'sul de minas - carmo de minas' 'natou county' 'nyeri' 'eastern uganda'
 'tolima' 'kiambu' 'sipi, mt elgon' 'nuevo oriente' 'eastern' 'huila'
 'boquete' 'acatenango' 'addis ababa' 'cajamarca'
 'eastern highlands province' 'apaneca'
 'ataco, apaneca - ilamatepec mountain range' 'kirinyaga'
 'bulambuli eastern region' 'huehuetenango' 'kapchorwa' 'west valley'
 'central kenya' 'oriente' 'santander' 'lintong'
 'kefa zone, gimbo distict, at a place called woka araba, south west ethiopia.'
 'pasto' 'aricha' 'cundinamarca' 'tarrazu' 'veracruz' 'grama valley'
 'mexico' 'chuva, san marcos' 'mbeya' 'coatepec'
 'dongshan dist., tainan city 臺南市東山區' 'thailand' 'marcala' 'santa ana'

In [7]:
# What are the different colours of coffee beans included in the dataset?
coffee_df['color'].unique()

array(['Green', nan, 'Bluish-Green', 'None', 'Blue-Green'], dtype=object)

In [8]:
# Splitting datasets to train and test datasets
coffee_train_df, coffee_test_df = train_test_split(coffee_df, test_size=0.2, random_state=123)

In [9]:
print(coffee_train_df.shape)
print(coffee_test_df.shape)

(1048, 43)
(263, 43)


In [10]:
# Explore numerical features in the data set
coffee_train_df.describe()

total_cup_points  number_of_bags        aroma       flavor  \
count       1048.000000     1048.000000  1048.000000  1048.000000   
mean          82.178588      156.144084     7.559666     7.519179   
std            2.542936      130.775664     0.317137     0.340838   
min           63.080000        1.000000     5.080000     6.080000   
25%           81.080000       15.000000     7.420000     7.330000   
50%           82.500000      200.000000     7.580000     7.500000   
75%           83.580000      275.000000     7.750000     7.750000   
max           90.580000     1062.000000     8.750000     8.830000   

        aftertaste      acidity         body      balance  uniformity  \
count  1048.000000  1048.000000  1048.000000  1048.000000  1048.00000   
mean      7.399714     7.531097     7.522338     7.517948     9.84124   
std       0.352207     0.317110     0.289447     0.350441     0.48868   
min       6.170000     6.080000     6.330000     6.080000     6.00000   
25%       7.170000     7.330000     7.330000     7.330000    10.00000   
50%       7.420000     7.500000     7.500000     7.500000    10.00000   
75%       7.580000     7.750000     7.670000     7.750000    10.00000   
max       8.670000     8.750000     8.580000     8.750000    10.00000   

         clean_cup    sweetness  cupper_points     moisture  \
count  1048.000000  1048.000000    1048.000000  1048.000000   
mean      9.869513     9.922290       7.495363     0.088702   
std       0.547691     0.380899       0.435170     0.048210   
min       5.330000     6.000000       5.170000     0.000000   
25%      10.000000    10.000000       7.250000     0.090000   
50%      10.000000    10.000000       7.500000     0.110000   
75%      10.000000    10.000000       7.750000     0.120000   
max      10.000000    10.000000      10.000000     0.280000   

       category_one_defects      quakers  category_two_defects  \
count           1048.000000  1047.000000           1048.000000   
mean               0.458969     0.175740              3.582061   
std                1.990962     0.862244              5.443900   
min                0.000000     0.000000              0.000000   
25%                0.000000     0.000000              0.000000   
50%                0.000000     0.000000              2.000000   
75%                0.000000     0.000000              4.000000   
max               31.000000    11.000000             55.000000   

       altitude_low_meters  altitude_high_meters  altitude_mean_meters  
count           858.000000            858.000000            858.000000  
mean           1532.257826           1580.123429           1556.190628  
std            6473.547509           6472.893700           6472.806952  
min               1.000000              1.000000              1.000000  
25%            1100.000000           1100.000000           1100.000000  
50%            1300.000000           1320.000000           1310.640000  
75%            1600.000000           1650.000000           1600.000000  
max          190164.000000         190164.000000         190164.000000

In [11]:
#cleaning dataset
coffee_train_df_clean = coffee_train_df[coffee_train_df['color'].notna()]
coffee_train_df_cleaned = coffee_train_df_clean.query("color != 'None'")

In [12]:
# One exploratory question we wanted to explore was to see if the colour 
# of the coffee bean affect the total quality rating?

In [13]:
coffee_color_boxplot = alt.Chart(coffee_train_df_cleaned).mark_boxplot().encode(
    x = alt.X('total_cup_points', title = "Total quality rating (0-100 scale)"),
    y = alt.Y('color', title = "Colour of coffee beans"))
coffee_color_boxplot_mean = coffee_color_boxplot + coffee_color_boxplot.mark_circle(fill='black',size=11).encode(x='mean(total_cup_points)')
coffee_color_boxplot_mean

alt.LayerChart(...)

In [14]:
coffee_color_boxplot_mean.save("coffee_quality_eda_images/coffeebean_color_rating.svg")

In [15]:
# Another exploratory question we wanted to explore was to see if the country of origin
# of the coffee bean affect the total quality rating?
coffee_train_df_clean = coffee_train_df[coffee_train_df['country_of_origin'].notna()]

In [16]:
coffee_coo_boxplot = alt.Chart(coffee_train_df_clean).mark_boxplot().encode(
    x = alt.X('total_cup_points', scale = alt.Scale(domain=(55, 100)), 
              title = "Total quality rating (0-100 scale)"),
    y = alt.Y('country_of_origin', title = "Country of Origin"))
coffee_coo_boxplot_mean = (coffee_coo_boxplot 
                           + coffee_coo_boxplot.mark_circle(fill='black',size=11).encode(x='mean(total_cup_points)'))
coffee_coo_boxplot_mean

alt.LayerChart(...)

In [17]:
coffee_coo_boxplot_mean.save("coffee_quality_eda_images/coffee_coo_boxplot_mean.svg")

In [18]:
# We may want to investigate the relationships between numerical features to better 
# understand which features might be correlated to the target.

numeric_cols = list(coffee_train_df.select_dtypes(exclude='object').drop(columns =['altitude_low_meters', 'altitude_high_meters', 'aroma', 'flavor', 'aftertaste', 'acidity', 'body', 'balance', 'uniformity', 'sweetness', 'clean_cup', 'cupper_points']))
splom = (
    alt.Chart(coffee_train_df)
    .mark_point(opacity = 0.5, size = 0.5)
    .encode(
        x = alt.X(alt.repeat("repeat"), type = "quantitative", scale = alt.Scale(zero =  False)),
        y = alt.Y("total_cup_points", scale = alt.Scale(zero = False))
    )
    .properties(height = 80, width = 80)
    .repeat(numeric_cols, columns = 3)
    .configure_axis(labelFontSize = 11)
)

splom

alt.RepeatChart(...)

In [19]:
splom.save("coffee_quality_eda_images/numerical_features_corr.svg")

At this point of the exploratory data analysis project, we have only explored the relationship of numerical features and the target, `total_cup_points`. We plan to do further cleaning and preparation of dataset to explore relationships of other categorical features in this dataset as well. 